SG14/inplace_function.h

```diff
@@ -150,6 +150,76 @@ struct is_valid_inplace_dst : std::true_type
     );
 };
 
+// C++11 MSVC compatible implementation of std::is_invocable_r,
+// std version doesn't seem to respect return types and requires C++17.
+// See https://godbolt.org/z/T0bd3K for defect example.
+
+template<
+    typename Default,
+    template<typename...> class Expression,
+    typename... Ts
+>
+class detected_or
+{
+private:
+    template<
+        template<typename...> class Alias,
+        typename... Args,
+        typename Ret_t = Alias<Args...>
+    >
+    static auto ov(int ph) -> Ret_t;
+
+    template<
+        template<typename...> class Alias,
+        typename... Args
+    >
+    static Default ov(...);
+
+public:
+    using type = decltype(ov<Expression, Ts...>(0));
+};
+
+template<
+    typename R,
+    typename C,
+    typename... Args
+>
+using def_closure_return_type_convertible = std::is_convertible<
+    typename std::decay<R>::type,
+    decltype(std::declval<C>()(std::declval<Args>()...))
+>;
```

Comment:
This feels overly complicated. What if it were just something like (UNTESTED CODE):
```
template<class R> void accept(R);

template<class, class R, class F, class... Args> struct is_invocable_r_impl : std::false_type {};
template<class F, class... Args> struct is_invocable_r_impl<decltype(declval<F>()(declval<Args>()...), void()), void, F, Args...> : std::true_type {};
template<class F, class... Args> struct is_invocable_r_impl<decltype(declval<F>()(declval<Args>()...), void()), const void, F, Args...> : std::true_type {};
template<class R, class F, class... Args> struct is_invocable_r_impl<decltype(accept<R>(declval<F>()(declval<Args>()...))), R, F, Args...> : std::true_type {};

template<class R, class F, class... Args> using is_invocable_r = is_invocable_r_impl<void, R, F, Args...>;
```

Reply:
Whether that's simpler I guess depends on your tmp style, this avoids the use of `std::is_convertible` which fails with incomplete types on mac. So I'll try this.